xline/src/request_validation.rs

```diff
@@ -0,0 +1,414 @@
+const DEFAULT_MAX_TXN_OPS: usize = 128;
+
+/// Trait for request validation
+pub(crate) trait RequestValidator {
+    /// Validate the request
+    fn validation(&self) -> Result<(), ValidationError>;
+}
+
+impl RequestValidator for RangeRequest {
+    fn validation(&self) -> Result<(), ValidationError> {
+        if self.key.is_empty() {
+            return Err(ValidationError::new("key is not provided"));
+        }
+        if !SortOrder::is_valid(self.sort_order) || !SortTarget::is_valid(self.sort_target) {
+            return Err(ValidationError::new("invalid sort option"));
+        }
+
+        Ok(())
+    }
+}
+
+impl RequestValidator for PutRequest {
+    fn validation(&self) -> Result<(), ValidationError> {
+        if self.key.is_empty() {
+            return Err(ValidationError::new("key is not provided"));
+        }
+        if self.ignore_value && !self.value.is_empty() {
+            return Err(ValidationError::new(
+                "ignore value is set but value is provided",
+            ));
+        }
+        if self.ignore_lease && self.lease != 0 {
+            return Err(ValidationError::new(
+                "ignore lease is set but lease is provided",
+            ));
+        }
+
+        Ok(())
+    }
+}
+
+impl RequestValidator for DeleteRangeRequest {
+    fn validation(&self) -> Result<(), ValidationError> {
+        if self.key.is_empty() {
+            return Err(ValidationError::new("key is not provided"));
+        }
+
+        Ok(())
+    }
+}
+
+impl RequestValidator for TxnRequest {
+    fn validation(&self) -> Result<(), ValidationError> {
+        let opc = self
+            .compare
+            .len()
+            .max(self.success.len())
+            .max(self.failure.len());
+        if opc > DEFAULT_MAX_TXN_OPS {
+            return Err(ValidationError::new("too many operations in txn request"));
+        }
+        for c in &self.compare {
+            if c.key.is_empty() {
+                return Err(ValidationError::new("key is not provided"));
+            }
+        }
+        for op in self.success.iter().chain(self.failure.iter()) {
+            if let Some(ref request) = op.request {
+                match *request {
+                    Request::RequestRange(ref r) => r.validation(),
+                    Request::RequestPut(ref r) => r.validation(),
+                    Request::RequestDeleteRange(ref r) => r.validation(),
+                    Request::RequestTxn(ref r) => r.validation(),
+                }?;
+            } else {
+                return Err(ValidationError::new("request not provided in operation"));
+            }
+        }
+
+        let _ignore_success = check_intervals(&self.success)?;
+        let _ignore_failure = check_intervals(&self.failure)?;
+
+        Ok(())
+    }
+}
+
+/// Check if puts and deletes overlap
+fn check_intervals(ops: &[RequestOp]) -> Result<(HashSet<&[u8]>, Vec<KeyRange>), ValidationError> {
+    // TODO: use interval tree is better?
```

Comment:
Open an issue to track it?

Reply:
Is this issue #80 ? @themanforfree

Reply:
No, we didn't use interval tree now, There was no suitable rust implementation for interval trees when I write this part, so I leave a `TODO` here. and I'm not sure that using this interval tree will be better than the current implementation.

Reply:
I opened this in #409